### PR TITLE
Feature/image blocking

### PR DESCRIPTION
**Pull request by CH3RVB**
This is my first PR to here, so be nice to me /hj

This is my third (i think) crack at making something like this. Everything else was either too clunky or felt... annoying to use? So this is my final solution. LOL. 

Adds a modular image blocking system for users. Users can curate their own blocklist by blocking a specific model + ID. Blocked models + IDs have their image blurred out until they become unblocked again by the user. I say model + ID because right now it's an image only thing, but theoretically you could apply this in other ways? 

With a small dosage of Moif magic, it automatically applies to all mentions of the blocked image across the site, as long as they hotlink directly to the image URL. Very easy to use across exts, simply plug in the block widget, block an image, and go.

I'm unsure about some aspects, such as the route for the image blocking, as well as how many different world/other entries I should even put this thing on. Maybe the button is also too annoying to look at? Dunno. Either way, it's here :')

### CODE REVIEW

File: app/Helpers/Helpers.php [hunk 1]
```diff
@@ -444,8 +462,33 @@ function prettyProfileName($url) {
...
+
+    $block = App\Models\ImageBlock::where([
+        ['item_id', '=', $item->id],
+        ['item_type', '=', $model],
+        ['user_id', '=', $user->id],
+    ])->first();
+
+    if ($block) {
+        return true;
+    } else {
```
Comment (itinerare) [inline app/Helpers/Helpers.php:491]: Generally speaking, with returns, you don't need to encase the second in the if/else-- nothing beyond a  `return` is executed  to begin with, so you can simply have the first in the if, then the second freestanding.

File: app/Models/ImageBlock.php [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+class ImageBlock extends Model
+{
+
+    /**
+     * The attributes that are mass assignable.
+     *
+     * @var array
+     */
+    protected $fillable = [
+        'item_id', 'item_type', 'user_id'
```
Comment (itinerare) [inline app/Models/ImageBlock.php:17]: Referring to non-specific objects as "items" when "item" is a concept already extant in the codebase is... very confusing.

File: resources/views/layouts/app.blade.php [hunk 3]
```diff
@@ -82,6 +82,18 @@
     @endif
 
     @include('feed::links')
+
+    {{-- Image Blocking --}}
+    @if (Auth::check() && count(Auth::user()->blockedImages))
+        <style>
+            @foreach (Auth::user()->blockedImages as $image)
+                [src="{{ $image->item->imageUrl ? $image->item->imageUrl : ($image->item->currencyImageUrl ? $image->item->currencyImageUrl : ($image->item->avatarUrl ? $image->item->avatarUrl : '')) }}"] {
```
Comment (itinerare) [inline resources/views/layouts/app.blade.php:90]: This logic would be better done in an attribute on the image block model, I think. Would be a good use case for a switch statement, potentially.

File: resources/views/layouts/app.blade.php [hunk 4]
```diff
@@ -82,6 +82,18 @@
     @endif
 
     @include('feed::links')
+
+    {{-- Image Blocking --}}
+    @if (Auth::check() && count(Auth::user()->blockedImages))
```
Comment (itinerare) [inline resources/views/layouts/app.blade.php:87]: The count here should just be able to be `Auth::user()->blockedImages->count()`, since iirc what you get out of a relation is either directly or functionally a collection.

File: resources/views/widgets/_item_block.blade.php [hunk 5]
```diff
@@ -0,0 +1,7 @@
+ {!! Form::open(['url' => 'account/image-blocks/block/' . $item->id]) !!}
+ @php
+     $model = get_class($item);
+ @endphp
+ <input type="hidden" name="item_type" value="{{ $model }}" />
+ {!! Form::submit(checkItemBlock($item, Auth::user()) ? 'Unblock' : 'Block', ['class' => 'btn btn-success btn-sm '.(isset($float) && $float ? 'float-right' : '')]) !!}
```
Comment (itinerare) [inline resources/views/widgets/_item_block.blade.php:6]: Probably this should be clearer about *what's* being blocked.

File: resources/views/widgets/_item_block.blade.php [hunk 6]
```diff
@@ -0,0 +1,7 @@
+ {!! Form::open(['url' => 'account/image-blocks/block/' . $item->id]) !!}
+ @php
+     $model = get_class($item);
+ @endphp
+ <input type="hidden" name="item_type" value="{{ $model }}" />
```
Comment (itinerare) [inline resources/views/widgets/_item_block.blade.php:5]: While the risk is low in a project like this that's *already* open source, it's generally not the best idea to plunk stuff like model names out in the open; comments already went through this, so have a look [at the form for those](/corowne/lorekeeper/blob/develop/resources/views/comments/_form.blade.php), etc. for an idea as to what to do about it.

File: routes/lorekeeper/members.php [hunk 7]
```diff
@@ -52,6 +52,8 @@
     Route::post('bookmarks/edit/{id}', 'BookmarkController@postCreateEditBookmark');
     Route::get('bookmarks/delete/{id}', 'BookmarkController@getDeleteBookmark');
     Route::post('bookmarks/delete/{id}', 'BookmarkController@postDeleteBookmark');
+
+    Route::post('image-blocks/block/{id}', 'AccountController@postBlockUnblockImage');
```
Comment (itinerare) [inline routes/lorekeeper/members.php:56]: If you go a similar route to comments (wrt the above), this'll probably need to be changed accordingly, but the general gist and placement of the route are fine.

File: app/Services/UserService.php [hunk 8]
```diff
@@ -643,4 +661,47 @@ public function reactivate($user, $staff = null) {
...
+        DB::beginTransaction();
+
+        try {
+            $model = $data['item_type'];
+            $item = $model::find($id);
+            $user = Auth::user();
+
+            //check it's not already blocked, if not, make the block
+
+            $block = ImageBlock::where([
```
Comment (itinerare) [inline app/Services/UserService.php:684]: You could probably trim off a little work and express this as `$user->blockedImages->where([non-user-id-conditions etc etc])->first()` too!

File: app/Services/UserService.php [hunk 9]
```diff
@@ -643,4 +661,47 @@ public function reactivate($user, $staff = null) {
...
+     *
+     * @return bool
+     */
+    public function blockUnblockImage($data, $id)
+    {
+        DB::beginTransaction();
+
+        try {
+            $model = $data['item_type'];
+            $item = $model::find($id);
```
Comment (itinerare) [inline app/Services/UserService.php:679]: It might be worthwhile to double-check that the object exists just to be safe, since it's technically in the 'accepting semi-arbitrary input from users' zone.

**Review by itinerare — CHANGES_REQUESTED** [review]
A good stab at an arguably relatively complex task!

**Comment by itinerare** [comment]
Belatedly, also... it might seem counterintuitive, but it might be an idea to have a page (akin to bookmarks) with the current user's blocked images, as a ready point of reference.

File: app/Helpers/Helpers.php
Comment (Ne-wt) [inline app/Helpers/Helpers.php:1]: nit: run composer lint to stop these weird style changes

File: app/Helpers/Helpers.php (hunk 1 above)
Comment (Ne-wt) [reply app/Helpers/Helpers.php:491]: +1, can even just 
`return $block ? true : false`

File: app/Http/Controllers/Users/AccountController.php
Comment (Ne-wt) [inline app/Http/Controllers/Users/AccountController.php:1]: again with composer lint

File: app/Http/Controllers/Users/AccountController.php [hunk 10]
```diff
@@ -475,4 +504,26 @@ public function postReactivate(Request $request, UserService $service) {
...
+     *
+     */
+    public function postBlockUnblockImage(Request $request, UserService $service, $id)
+    {
+        $data = $request->only([
+            'item_type'
+        ]);
+
+        if ($service->blockUnblockImage($data, $id)) {
+
```
Comment (Ne-wt) [inline app/Http/Controllers/Users/AccountController.php:519]: should this be blank? if yes, change it to `!$service...` and remove else block

File: app/Models/ImageBlock.php (hunk 2 above)
Comment (Ne-wt) [reply app/Models/ImageBlock.php:17]: +1, recommend `block_object_id` or `image_id` `image_type` etc

File: app/Models/ImageBlock.php [hunk 11]
```diff
@@ -0,0 +1,51 @@
+<?php
+
+namespace App\Models;
```
Comment (Ne-wt) [inline app/Models/ImageBlock.php:3]: should this be in the user space, since it is user specific, and be `UserImageBlock`? imo I think it should be since there is no user agnostic rows

File: app/Services/UserService.php
Comment (Ne-wt) [inline app/Services/UserService.php:1]: +1 to lint

File: database/migrations/2024_04_25_052128_add_image_blocking.php [hunk 12]
```diff
@@ -0,0 +1,30 @@
+<?php
+
+use Illuminate\Database\Migrations\Migration;
+use Illuminate\Database\Schema\Blueprint;
+use Illuminate\Support\Facades\Schema;
+
+return new class extends Migration
```
Comment (Ne-wt) [inline database/migrations/2024_04_25_052128_add_image_blocking.php:7]: this should be `class FileName extends Migration` to stop errors on older installs

File: resources/views/user/_profile_content.blade.php [hunk 13]
```diff
@@ -6,6 +6,9 @@
     <div class="col-md-2 text-center">
         <!-- User Icon -->
         <img src="{{ $user->avatarUrl }}" class="img-fluid rounded-circle" style="max-height: 125px;" alt="{{ $user->name }}'s Avatar">
+        @if(Auth::check())
+                @include('widgets._item_block', ['item' => $user])
+            @endif
```
Comment (Ne-wt) [inline resources/views/user/_profile_content.blade.php:11]: nit: indents

Comment (Ne-wt) [reply resources/views/user/_profile_content.blade.php:11]: double nit: item should not be the variable name

File: resources/views/widgets/_item_block.blade.php [hunk 14]
```diff
@@ -0,0 +1,7 @@
+ {!! Form::open(['url' => 'account/image-blocks/block/' . $item->id]) !!}
+ @php
+     $model = get_class($item);
```
Comment (Ne-wt) [inline resources/views/widgets/_item_block.blade.php:3]: further on prior nit: `$item` is not the best variable name

File: resources/views/world/_currency_entry.blade.php [hunk 15]
```diff
@@ -8,6 +8,9 @@
         <h3>{{ $currency->name }} @if ($currency->abbreviation)
                 ({{ $currency->abbreviation }})
             @endif
+            @if(Auth::check())
+                @include('widgets._item_block', ['item' => $currency, 'float' => true])
```
Comment (Ne-wt) [inline resources/views/world/_currency_entry.blade.php:12]: float seems to be true in most places, consider it being default to true and explicitly pass it as false, but that's mostly preference

**Review by ScuffedNewt — CHANGES_REQUESTED** [review]
mostly nits, agree with what mercury has said
apologies if comments are short - in work PR review mode

**Comment by Ne-wt** [comment]
Also unrelated, I find this a little bit strange to include in the default LK install, considering things like toyhouse etc etc don't employ this on general images, and if a specific lorekeeper has such images on items etc that may need to be blocked its best installed individually, imo
@itinerare thoughts?

**Comment by itinerare** [comment]
On some level, I agree... but users may need to block seemingly innocuous stuff in certain cases, so it's not necessarily sufficient to install it on a case-by-case basis/only where more common points of friction occur.

**Comment by itinerare** [comment]
To some degree, I feel like if it's worth including in core at all then there's no sense in undercutting it by putting it behind a toggle... 

I think the other part of it is that TH has a different use case than we do, so it's not necessarily practical to make a direct comparison there; TH doesn't really have images that persistently exist around the site in the same way (items, etc.) to be an issue in the first place. Generally speaking, you have more leeway to simply not look at a character or image on there that causes issues, and iirc can block characters-- so it does have _some_ manner of comparable function, just that the context is different.
The more direct analogue would probably be pet sites-- and imo the fact that those generally don't have this kind of option isn't in their favor.

It's one of those things where it's the safer and more considerate thing to be proactive about it and spare folks unnecessary stress, even if it feels like "overkill" in a lot of cases/to include in core.

**Comment by Ne-wt** [comment]
> The more direct analogue would probably be pet sites

definitely, hit the nail on the head! brain wasn't braining yday

> It's one of those things where it's the safer and more considerate thing to be proactive about it and spare folks unnecessary stress, even if it feels like "overkill" in a lot of cases/to include in core.

firmly in the camp of <this is probably not a feature conducive? to core> since most items etc should not be causing these issues ideally, especially when compared to Lorekeepers mainstream counterparts
That being said, I think this would be better placed as having `Characters` as default -- since when #876 is inevitably cleaned and finished, users may still find that they want to block a certain character even if does is not necessarily fall under a "warning needed" category. 
In fact, I think that would be a much more requested feature than items

TLDR; IMO, if this feature is to be added...

- Change the primary focus of the default configuration to be for characters
- A toggle, if not (since other use cases would need to be manually added by a developer)

File: app/Models/ImageBlock.php [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+    protected $fillable = [
+        'item_id', 'item_type', 'user_id'
+    ];
+
+    /**
+     * The table associated with the model.
+     *
+     * @var string
+     */
+    protected $table = 'image_blocks';
```
Comment (Ne-wt) [inline app/Models/ImageBlock.php:25]: following up on my prior comment about user space too, ideally the table should be `user_image_blocks` in similar fashion

**Comment by itinerare** [comment]
> definitely, hit the nail on the head! brain wasn't braining yday

Haha, all good, been there.

Anyhow, yeah, I'm inclined to agree that characters are the bigger/more probable use case, so I think it's sensible enough to aim this more firmly _at_ that use case-- with support for other things essentially as a side benefit (albeit one that I do still find in favor of, personally). I'd agree that most items, etc. _shouldn't_ cause issues/it's generally good to avoid making ones that are liable to, but I'd rather provide users the ability to address these things for themselves.

File: app/Helpers/Helpers.php
Comment (itinerare) [reply app/Helpers/Helpers.php:1]: Probably I'm just going to switch it to running on push for all branches again so neither reviewers nor contributors need to think about it, since my clever idea wrt running on PRs only was stymied (fairly reasonably) by repo perms... It was worth a shot |D

Comment (Ne-wt) [reply app/Helpers/Helpers.php:1]: 👌 
> It was worth a shot |D

can't work out all the time!

File: app/Helpers/Helpers.php (hunk 1 above)
Comment (Ne-wt) [reply app/Helpers/Helpers.php:491]: actually follow up on this i'd do

`function checkImageBlock($image, $user)
{
    return App\Models\ImageBlock::where([
        ['image_id', '=', $image->id],
        ['image_type', '=', get_class($image)],
        ['user_id', '=', $user->id],
    ])->exists();
}`

File: app/Helpers/Helpers.php
Comment (itinerare) [reply app/Helpers/Helpers.php:1]: Did that while I was thinking about it-- so it _should_ now be as easy as just merging in latest develop. Hopefully.

File: app/Models/ImageBlock.php (hunk 11 above)
Comment (itinerare) [reply app/Models/ImageBlock.php:3]: Good call yeah, inclined to agree

File: database/migrations/2024_04_25_052128_add_image_blocking.php (hunk 12 above)
Comment (itinerare) [reply database/migrations/2024_04_25_052128_add_image_blocking.php:7]: _Does_ this cause issues? My understanding was that upstream (wrt Laravel) this change was following PHP supporting it, so it doesn't make much sense to me that (assuming PHP is upgraded, which needs to happen regardless) this should be a problem.

File: app/Models/ImageBlock.php (hunk 16 above)
Comment (itinerare) [reply app/Models/ImageBlock.php:25]: Likewise, agree on this

File: database/migrations/2024_04_25_052128_add_image_blocking.php (hunk 12 above)
Comment (Ne-wt) [reply database/migrations/2024_04_25_052128_add_image_blocking.php:7]: I've found on SOME sites it can be an issue, & quiet often on local (people on windows / older xampp)
perhaps in this case it *should* be fine since this is on v3 which should implicitly mean that the php version is up to par, so take this comment with a grain of salt

Comment (itinerare) [reply database/migrations/2024_04_25_052128_add_image_blocking.php:7]: I'm inclined to let this be since yeah, v3 explicitly requires PHP be reasonably up-to-date (whereas on current release it's supported but not technically _forced_)...

**Comment by CH3RVB** [comment]
cancelling due to feedback
